pkg/plugins/pluginrequestmeta/plugin_request_meta.go

```diff
@@ -0,0 +1,45 @@
+package pluginrequestmeta
+
+import (
+	"context"
+	"errors"
+)
+
+// StatusSource is an enum-like string value representing the source of a
+// plugin query data response status code
+type StatusSource string
+
+const (
+	StatusSourceNone       StatusSource = ""
+	StatusSourcePlugin     StatusSource = "plugin"
+	StatusSourceDownstream StatusSource = "downstream"
+)
+
+type statusSourceCtxKey struct{}
+
+// StatusSourceFromContext returns the plugin request status source stored in the context.
+// If no plugin request status source is stored in the context, [StatusSourceNone] is returned.
+func StatusSourceFromContext(ctx context.Context) StatusSource {
+	value, ok := ctx.Value(statusSourceCtxKey{}).(*StatusSource)
+	if ok {
+		return *value
+	}
+	return StatusSourceNone
+}
+
+// WithStatusSource sets the plugin request status source for the context.
+func WithStatusSource(ctx context.Context, s StatusSource) context.Context {
+	return context.WithValue(ctx, statusSourceCtxKey{}, &s)
+}
+
+// WithDownstreamStatusSource mutates the provided context by setting the plugin request status source to
+// StatusSourceDownstream. If the provided context does not have a plugin request status source, the context
+// will not be mutated. This means that [WithStatusSource] has to be called before this function.
+func WithDownstreamStatusSource(ctx context.Context) error {
```

Comment:
Hmm why is this different compared to withStatusSource?

Reply:
This is indeed a bit confusing.

`WithStatusSource` returns a new `context.Context` with the new value, so the caller will have to pass it to the function it calls.

`WithDownstreamStatusSource` mutates the provided `context.Context`, so you can technically change the context that the caller passed to a function, and the caller will see it.

I did it this way because we mutate the context inside the function called here:

https://github.com/grafana/grafana/blob/ee7b4985ef18126d299292286dbe9d3eb1fb8caa/pkg/services/pluginsintegration/clientmiddleware/metrics_middleware.go#L122

(the context is mutated in the wrapped function, here):

https://github.com/grafana/grafana/blob/ee7b4985ef18126d299292286dbe9d3eb1fb8caa/pkg/services/pluginsintegration/clientmiddleware/metrics_middleware.go#L197